aws/rust-runtime/aws-runtime/src/profile/profile_file.rs

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Config structs to programmatically customize the profile files that get loaded
+
+use std::fmt;
+use std::path::PathBuf;
+
+/// Provides the ability to programmatically override the profile files that get loaded by the SDK.
+///
+/// The [`Default`] for `ProfileFiles` includes the default SDK config and credential files located in
+/// `~/.aws/config` and `~/.aws/credentials` respectively.
+///
+/// Any number of config and credential files may be added to the `ProfileFiles` file set, with the
+/// only requirement being that there is at least one of them. Custom file locations that are added
+/// will produce errors if they don't exist, while the default config/credentials files paths are
+/// allowed to not exist even if they're included.
+///
+/// # Example: Using a custom profile file path
+///
+/// ```no_run,ignore
+/// use aws_runtime::profile::profile_file::{ProfileFiles, ProfileFileKind};
+/// use std::sync::Arc;
+///
+/// # async fn example() {
+/// let profile_files = ProfileFiles::builder()
+///     .with_file(ProfileFileKind::Credentials, "some/path/to/credentials-file")
+///     .build();
+/// let sdk_config = aws_config::from_env()
+///     .profile_files(profile_files)
+///     .load()
+///     .await;
+/// # }
+/// ```
+#[derive(Clone, Debug)]
+pub struct ProfileFiles {
+    pub(crate) files: Vec<ProfileFile>,
+}
+
+impl ProfileFiles {
+    /// Returns a builder to create `ProfileFiles`
+    pub fn builder() -> Builder {
+        Builder::new()
+    }
+}
+
+impl Default for ProfileFiles {
+    fn default() -> Self {
+        Self {
+            files: vec![
+                ProfileFile::Default(ProfileFileKind::Config),
+                ProfileFile::Default(ProfileFileKind::Credentials),
+            ],
+        }
+    }
+}
+
+/// Profile file type (config or credentials)
+#[derive(Copy, Clone, Debug)]
+pub enum ProfileFileKind {
+    /// The SDK config file that typically resides in `~/.aws/config`
+    Config,
+    /// The SDK credentials file that typically resides in `~/.aws/credentials`
+    Credentials,
+}
+
+impl ProfileFileKind {
+    pub(crate) fn default_path(&self) -> &'static str {
+        match &self {
+            ProfileFileKind::Credentials => "~/.aws/credentials",
+            ProfileFileKind::Config => "~/.aws/config",
+        }
+    }
+
+    pub(crate) fn override_environment_variable(&self) -> &'static str {
+        match &self {
+            ProfileFileKind::Config => "AWS_CONFIG_FILE",
+            ProfileFileKind::Credentials => "AWS_SHARED_CREDENTIALS_FILE",
+        }
+    }
+}
+
+/// A single profile file within a [`ProfileFiles`] file set.
+#[derive(Clone)]
+pub(crate) enum ProfileFile {
```

Comment:
I did a big rename of these and related types. Let me know what you think.

For ease of review, I kept using the same type names as the old ones in `aws-config` by using the deprecated type aliases. I can switch them to use the new names, I just thought this might make it easier to understand what's actually changed and what's basically the same after the move+rename